services/compute/baremetalmachine/baremetalmachine.go

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache v2.0 License.
+
+package baremetalmachine
+
+import (
+	"github.com/microsoft/moc/pkg/convert"
+	"github.com/microsoft/moc/pkg/errors"
+
+	"github.com/microsoft/moc-sdk-for-go/services/compute"
+
+	"github.com/microsoft/moc/pkg/status"
+	wssdcloudcompute "github.com/microsoft/moc/rpc/cloudagent/compute"
+)
+
+func (c *client) getWssdBareMetalMachine(bmm *compute.BareMetalMachine, group string) (*wssdcloudcompute.BareMetalMachine, error) {
+	if bmm.Name == nil {
+		return nil, errors.Wrapf(errors.InvalidInput, "Bare Metal Machine name is missing")
+	}
+	if len(group) == 0 {
+		return nil, errors.Wrapf(errors.InvalidGroup, "Group not specified")
+	}
+	storageConfig, err := c.getWssdBareMetalMachineStorageConfiguration(bmm.StorageProfile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get Storage Configuration")
+	}
+	hardwareConfig, err := c.getWssdBareMetalMachineHardwareConfiguration(bmm)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get Hardware Configuration")
+	}
+	securityConfig, err := c.getWssdBareMetalMachineSecurityConfiguration(bmm)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get Security Configuration")
+	}
+	osconfig, err := c.getWssdBareMetalMachineOSConfiguration(bmm.OsProfile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get OS Configuration")
+	}
+
+	networkConfig, err := c.getWssdBareMetalMachineNetworkConfiguration(bmm.NetworkProfile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get Network Configuration")
+	}
+
+	bmmOut := wssdcloudcompute.BareMetalMachine{
```

Comment:
Are you not outputting nodeName on purpose?

Reply:
This is mostly cloned from VM where it was missing. Still learning, but this is _In_ path to the cloud agent where it may not be possible to know the node agent's name. I think it is something to be resolved in the cloud agent (currently work in progress) and set in the _Out_ path which we do (see your other comment).

Reply:
You are right, this should not be settable by the SDK.

Reply:
I may have caused confusion. In my proposed design, bare metal machines are created (the first time on-boarding) after the node creation. So at that time the setup script knows the node name and should pass it to the BMM. Otherwise the cloud agent does not know the node. For VMs, passing the node name is optional because the cloud agent picks a node based on VM size requirements, which I think makes sense. For bare metal, the node agent host is the BMM host. So, you were right, we should set the nodeName here.

Reply:
Interesting, I was actually thinking about it the other way around. We first have BMM in cloud agent. This is how cloud agent can talk to BMC for example (as defined in the BMM, eventually, provider specific). Then CA through the BMM provider provisions the BMM and deploys Node Agent. Only then there would be Node entity, which can be added into the BMM entity. Curious about your thoughts on this.

Reply:
How does BMM entity in CA talk to the BMM entity (or any entity) in node agent? I believe it is done via the node entity in CA. See GetNodeVirtualMachineClient() for VM reference. I was thinking I'd create a corresponding GetNodeBareMetalMachineClient(). In my mind the steps were following:

Via setup script at setup time:
1. Configure cloud agent
2. Configure node agent
3. Create location
4. Create group
5. Create node
6. Create BMM. This is on-boarding the BM node to the cloud. Only BM node capabilities are passed e.g., OS/storage/network info.

Via BM controller at a later time:
7. Update BMM. This is to add the BM node to a target cluster. Cloud-init (and possibly a target image ref) is passed.

Reply:
Step 2 expects that you have OS already deployed on that host, right? (Not for step 1, since Cloud Agent can be running elsewhere.) Here is my counter point scenario:
1) Say I have done what you proposed on a single host
2) I want to add another BM host; there is no OS, so first I need to onboard the host to MOC to let MOC drive the OS deployment
3) To do that, I create the BMM entity and pass BMC creds to it
4) MOC BMM entity creation is passed to the BMM provider, which then goes and deploys OS onto the host through BMC (no Node Agent yet), and initializes the host with cloud-init (this can deploy and configure Node Agent)
5) With Node Agent ready, additional operations can be performed through NodeAgent.

Now this is not what we are doing for March; for March, we will deploy OS upfront through Robin and so your flow will work. But I just wanted to point out that we might need to refactor that post March.

Here is my current thinking on the March flow: http://aka.ms/robin-march
Here is my current thinking on post March: https://aka.ms/robin-provider-no-dashboard

If you look at the second flow and the Add host to ECF section, it illustrates what I was trying to put in that counter point scenario. If you see anything wrong with the flows or have any other thoughts or feedback, would be happy to discuss and augment.

Reply:
I see - thanks. Yes, my steps assumed the OS is pre-deployed. The flow makes sense mostly. I have some questions, but will ask them privately.